### DISCUSSION TITLE
Input element with x-model does not update when data is updated

### DISCUSSION THREAD

**Discussion by peacememories** [alpinejs/alpine#868] [1. Help] [answered]
Hi, and thanks for creating this neat framework.
I've just started using it, and I've run into a small issue with resetting a form.
The form looks like this:

```html
<template x-data="passwordModal()" x-if="$store.modal.open == 'password'">
    <div id="changePasswordDialog" class="dialog">
      <div class="icon" x-on:click="$store.modal.open = false"><?php include 'img/cancel.svg'; ?></div>
      <h2>Passwort ändern</h2>
      <span id="changePasswordReply"></span>
      <form method="post" name="loginform" id="passwordForm">
        <div class="container">
          <label for="old_password">Altes Passwort:</label>
          <input id="old_password" type="text" name="old_password" x-model="passwordOld">
          <label for="new_password">Neues Passwort:</label>
          <input id="new_password" type="text" name="new_password" x-model="password">
          <label for="new_password_repeat">Neues Passwort wiederholen:</label>
          <input id="new_password_repeat" type="text" name="new_password_repeat" x-model="passwordRe">
        </div>
        <button name="changePassword" id="submit" x-on:click.prevent="submit">Absenden</button>
      </form>
    </div>
  </template>
```

with `passwordModal()` looking like this:

```js
function passwordModal() {
    return {
        passwordOld: "",
        password: "",
        passwordRe: "",
        submit() {
            this.reset()
        },
        reset() {
            console.log(this.password)
            this.password = ""
            this.passwordRe = ""
            this.passwordOld = ""
        }
    }
}
```

What I've already tried is using `reset` directly.
What I've also noticed is that the `Proxy` element I get in my console when i log from `reset` is correctly updated when I type in the fields, but when I change the values there they are not reflected in the interface.

It might have something to do with x-if, or I might be doing something else wrong, but I can't seem to find the problem.

Side note: the `$store` is from [Spruce](https://github.com/ryangjchandler/spruce)

**Comment by ryangjchandler** [comment, marked as answer]
Does the `x-model` functionality work correctly if you don't use a `<template>` tag and instead use a `<div>` without any conditions such as `x-if` or `x-show`?

**Reply by peacememories** [reply]
I just tested that, and it works if i remove the template and use an `x-show` on the div instead.
This works for this usecase, but I have another modal where I would ideally use `x-init` with `x-if` to initialize data when opening the modal, so I'd really like to make this work with `template` 😅 Also I would really like to understand why it doesn't work, just to understand Alpine better.

In any case, many thanks for the suggestion 😃

**Reply by ryangjchandler** [reply]
When you use `x-if`, the elements aren't actually in the DOM. That means that when they _are_ rendered, they actually end up outside of the `<template>` tag and therefore don't belong to a component and Alpine won't register all of the event listeners etc, because it doesn't walk through the `templateContent`.

Best thing to do is wrap is all in an `<div>` and perhaps use a `<template>` inside of that `<div>`.. can't say I've ever needed to use a `<template>` tag as my `x-data` host.

**Reply by peacememories** [reply]
So basically i should probably use a wrapping component when using `x-if`? Seems reasonable, thanks :)

**Reply by peacememories** [reply]
Can I just say that after years and years of elm, react, svelte, vue and so forth, alpine is a very welcome breath of fresh air. Especially since it gives me an actual way to tell people "hey, that manual innerHtml wrangling you're doing? I've got a solution for that"
Very happy so far 😄

**Reply by peacememories** [reply]
One more question if I may:
What is the best way to react to a modal being opened, for example to fetch the contents? I don't really want to do this on the button click handler that opens the modal, because that can be easily overlooked if the modal ever gets opened from somewhere else

**Reply by ryangjchandler** [reply]
You can use a watcher on a bit of data, such as an `open` prop.

**Reply by peacememories** [reply]
Problem is, the props are in Spruce. I could use `Spruce.watch` I'm just a bit worried that I might leave a bunch of Spruce watches lying around if the structure of the page changes. For this project it's probably not going to matter though, so I'll use that^^

Thanks for all your help!

**Reply by ryangjchandler** [reply]
Yeah, `Spruce.watch` could work too. Not really any other way to listen for changes, unless you dispatch an event but that can get messy too.
